### PR TITLE
feat: New Component SiteAlert

### LINKED ISSUE #981: New Component: Site Alert
# [Site Alert](https://designsystem.digital.gov/components/site-alert/)

## USWDS Version

2.10.0

## Description

When your agency has urgent information to share with the public, it’s essential that it stands out from other content on the site. A site alert prominently displays critical, time-sensitive warnings or directions across every page so that users see it whenever they visit the site. Using the same site alert treatment across agencies helps create a consistent and predictable way for users to find urgent information across all government websites.

## Requirements (proposed)

**Prop Interface**

```JSX
Props = {
  heading?: string // Default undefined (to allow Site alert with no header)
  variant: "info" | "emergency"
  showIcon?: boolean // Default true
  slim?: boolean // Default false
  className?: string
  children: React.ReactNode  
} & JSX.IntrinsicElements['section']
```

## Example Implementation (proposed)

```JSX
return (
  <SiteAlert heading="Short alert message">
    Additional context and followup information including <Link" href="#">a link</Link>.
  </SiteAlert>
)
```

**More Details**

- `<section>`root element
- The Alert component that we already have makes up most of the content of a SiteAlert. It's worth exploring if SiteAlert can just wrap an Alert and modify as necessary depending on the variant.

### PR DESCRIPTION
**Pull request by SirenaBorracha**
# Summary

This PR adds the [USWDS SiteAlert component](https://designsystem.digital.gov/components/site-alert). 

USWDS SiteAlert has four variants: `info`, `emergency`, `slim`, and `no-icon` that each apply different styles to the `<section>` root element. Brandon pointed out that these styles are not all mutually exclusive. 

In the pursuit of customizability, instead of including them all as possible values of the `variant` prop, two of the four are passed in as individual, optional properties: `showIcon` and `slim`. Separate props are preferred so that these styles can be combined with the two mutually exclusive variants, `info` and `emergency`, however the user wants.
 
The USWDS examples don't show all possible combinations, but I've included stories for the most likely combinations. In the last example, at Hana's suggestion, I used [Storybook Controls](https://storybook.js.org/docs/react/essentials/controls) to show the different applications  and combinations of styles with the `slim`, `showIcon`, and `variant` properties. If your control panel is hidden in Storybook, pressing the `D` key should make it reappear. Controls are only available for the last Storybook example.

## Related Issues or PRs

closes #981 

## How To Test

Check out this branch and run `yarn test` to execute tests and `yarn storybook` to see the new component in action. 

Alternatively, you can scroll to the bottom of this page and click "Show environments" on the right above the comment input box to access this component in Storybook. To see the different styles applied with Controls on the last story, press the `D` key in Storybook to access your controls panel and toggle the application of the `slim`, `showIcon`, and `variant` properties.
 

#### Standard SiteAlert, informational and emergency
Screen Shot 2021-04-08 at 5 10 19 PM

#### SiteAlert no header, informational and emergency
Screen Shot 2021-04-08 at 5 10 40 PM

#### SiteAlert with list, informational and emergency
Screen Shot 2021-04-08 at 5 11 16 PM

#### Slim Emergency Alert, Emergency Alert No Icon
Screen Shot 2021-04-08 at 5 12 55 PM

#### SiteAlert With Custom Controls (default values)
Screen Shot 2021-04-08 at 5 17 23 PM

#### SiteAlert With Custom Controls `slim`, `showIcon`, and `variant`
Screen Shot 2021-04-08 at 5 16 20 PM

### CODE REVIEW

File: src/components/SiteAlert/SiteAlert.stories.tsx
Comment (SirenaBorracha) [reply src/components/SiteAlert/SiteAlert.stories.tsx:61]: That's what I had initially, but then I swapped `Link` out since I figured they were just generic links 😅

**Review by haworku — COMMENTED** [review]
LGTM! I wanted to check a couple things though a bit outside the actual React component work here:
- @SirenaBorracha have you explored storybook controls at all? It might be something you would be interested in. We could very easily reduce our storybook stories and instead show the `slim` and `showIcon` differences with [Storybook controls](https://storybook.js.org/docs/react/essentials/controls). I'd be happy to pair with you in the future to check out how that would work. 
- @brandonlenz @suzubara  Didn't we say we want to start trying to make the uswds version bumps their own separate PR? Maybe this would be a good time.

File: src/components/SiteAlert/SiteAlert.test.tsx
Comment (brandonlenz) [reply src/components/SiteAlert/SiteAlert.test.tsx:85]: ooo and a quick follow up on these lines since I didn't notice on my first pass:

if you ever find yourself doing `getBy___` or `queryBy___` for the same thing multiple times in a test you can factor that out into a const, e.g:

```TSX
const heading = getByRole('heading')
expect(heading).toBeInTheDocument()
expect(heading).toHaveClass('usa-alert__heading')
```

I'm not sure if the performance impact of querying the dom for a specific element is high, but its an easy optimization we can do to keep the tests efficient.

**Comment by brandonlenz** [comment]
> @brandonlenz @suzubara Didn't we say we want to start trying to make the uswds version bumps their own separate PR? Maybe this would be a good time.

@haworku yep! We've got https://github.com/trussworks/react-uswds/issues/1097 for this effort. I got started on it today and hope to finish this week so we can get that in before these. Among other things it'll make the Happo diff for this PR just show the SiteAlert added renders. So we might get this to a point where its "approved" but hold off on merging to main. We have a couple of outstanding 2.9.0 PRs that _probably_ should also go through first as well*. 

###### * Unless we go with Option 2.A in  the [release & branching ADR](https://github.com/trussworks/react-uswds/pull/1026) in which case we just get things merge in whatever order they're ready 😉

File: src/components/SiteAlert/SiteAlert.stories.tsx
Comment (brandonlenz) [inline src/components/SiteAlert/SiteAlert.stories.tsx:29]: Swapping out `className="usa-link"` with `variant="unstyled"` will apply the `usa-link` class internally. I think we should go for that approach to encapsulate the uswds class.

File: src/components/SiteAlert/SiteAlert.stories.tsx
Comment (brandonlenz) [reply src/components/SiteAlert/SiteAlert.stories.tsx:79]: I like that approach! We might need to expand to accept `string | JSX.Element | JSX.Element[]` in the event of multiple child elements.

I think the only advantage of making the `<AlertText />` subcomponent is for the scenario where someone wanted to do something like:

```TSX
<SiteAlert>
  <AlertText>Here's some text about the list below:<AlertText />
  <ul className="usa-list">
    <li>something</li>
    <li>something else</li>
  </ul>
<SiteAlert>
```

The `string | JSX.Element | JSX.Element[]` approach doesn't preclude that so we could proceed that route and always leave `AlertText` as an enhancement should the need arise.

File: src/components/SiteAlert/SiteAlert.stories.tsx
Comment (brandonlenz) [reply src/components/SiteAlert/SiteAlert.stories.tsx]: Oh nice! Yea I guess I just have a slight preference against seeing a uswds class explicitly passed through a `className` prop. Sine the whole point of ReactUSWDS is to create a react wrapper for uswds, as soon as consumers have to manually write a uswds classname they're effectively using uswds directly and not relying on the reacty-ness of ReactUSWDS.

But there's definitely places where its appropriate or beyond the scope of a PR so I won't always call it out. Like `usa-list`. We use that in a few places and could totally create a `<List />` component to encapsulate that class, but I wouldn't introduce that for this PR.

**Comment by SirenaBorracha** [comment]
>... 
> @ SirenaBorracha have you explored storybook controls at all? It might be something you would be interested in. We could very easily reduce our storybook stories and instead show the `slim` and `showIcon` differences with [Storybook controls](https://storybook.js.org/docs/react/essentials/controls). I'd be happy to pair with you in the future to check out how that would work.
> ...

I'd prefer to remove some of those stories in this PR given the choice. 

I just spent ~30m trying to get controls working for `slim` and then `variant` following the instructions, but I'm having trouble. The panel doesn't show in Chrome because I closed it that one time, so instead I'm using Safari, but then even when I do see the panel in Safari there's an alert above it (lol meta) "*This story is not configured to handle controls*"  with a link to the Controls README. I just threw some time on @brandonlenz's calendar for tomorrow, but if he and I can't figure it out I'll put time on yours 🙃

File: src/components/SiteAlert/SiteAlert.stories.tsx
Comment (haworku) [reply src/components/SiteAlert/SiteAlert.stories.tsx:79]: > The string | JSX.Element | JSX.Element[] approach doesn't preclude that so we could proceed that route and always leave AlertText as an enhancement should the need arise.

Nice.

**Comment by haworku** [comment]
@SirenaBorracha  a couple things to check out re: controls.
- Try hitting `d` key to see if you can get  controls panel opening and closing. If not, I also saw once on a project I had to delete sb cache and rebuild storybook to get things working again because I got in a weird state.
- You may have to slightly adjust your story to get controls working using args. The [docs](https://storybook.js.org/docs/react/essentials/controls) have some examples.  Also you could try writing stories in the [CSF forma](2914#issuecomment-717289707)t instead of the outdated format we still use from before storybook 6. 
>  To use the Controls addon, you need to write your stories using args. Storybook will automatically generate UI controls based on your args and what it can infer about your component.

File: src/components/SiteAlert/SiteAlert.stories.tsx
Comment (brandonlenz) [reply src/components/SiteAlert/SiteAlert.stories.tsx:48]: Whoops, got this logic wrong. You shouldn't need to pass a variant or a className and it should apply `usa-link` internally.

**Review by brandonlenz — COMMENTED** [review]
Re-reviewing to indicate my approval! (We'll have to merge main and re-approve anyways after the 2.10.x, but this is for paper trail so I don't feel the need to do a completed review again)

**Review by brandonlenz — DISMISSED** [review]

**Review by brandonlenz — APPROVED** [review]
🥳

**Review by brandonlenz — CHANGES_REQUESTED** [review]
Oh I approved preemptively. Could you remove the empty whitespace added to package.json to this one as well? I'll eagerly approve after that!

File: yarn.lock
Comment (suzubara) [inline yarn.lock:13951]: where did this white space come from?

Comment (SirenaBorracha) [reply yarn.lock:13951]: I have no idea. Removing it.

**Review by brandonlenz — APPROVED** [review]

**Pull request merged**
